execinfra: prefer leased table descriptors

postamar opened the pull request:
This commit makes better use of the table descriptor protos in the
DistSQL specs by first checking if they've already been leased. If so,
then we use those instead of re-generating catalog.TableDescriptor
instances.

Release note: None

Commit:
execinfra: prefer leased table descriptors

This commit makes better use of the table descriptor protos in the
DistSQL specs by first checking if they've already been leased. If so,
then we use those instead of re-generating catalog.TableDescriptor
instances.

This has a statistically-significant impact on memory allocations, as
illustrated with this microbenchmark which I ran on my development
machine:

    name                    old time/op    new time/op    delta
    KV/Scan/SQL/rows=1-16      190µs ± 7%     184µs ± 4%   -3.60%  (p=0.016 n=10+8)
    KV/Scan/SQL/rows=10-16     196µs ± 4%     198µs ±12%     ~     (p=0.762 n=8+10)

    name                    old alloc/op   new alloc/op   delta
    KV/Scan/SQL/rows=1-16     19.5kB ± 1%    17.4kB ± 1%  -11.12%  (p=0.000 n=9+10)
    KV/Scan/SQL/rows=10-16    21.0kB ± 1%    18.9kB ± 1%  -10.20%  (p=0.000 n=10+10)

    name                    old allocs/op  new allocs/op  delta
    KV/Scan/SQL/rows=1-16        222 ± 0%       210 ± 1%   -5.59%  (p=0.000 n=7+10)
    KV/Scan/SQL/rows=10-16       256 ± 0%       244 ± 0%   -4.84%  (p=0.000 n=10+7)

This change opens us to the possibility of no longer shipping the whole
table descriptor proto in the DistSQL spec.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74722)
<!-- Reviewable:end -->

postamar commented on pkg/sql/execinfra/flow_context.go:
```diff
@@ -114,6 +122,40 @@ func (ctx *FlowCtx) Codec() keys.SQLCodec {
+	}
+	if ctx != nil && ctx.Descriptors != nil {
+		leased, _ := ctx.Descriptors.GetLeasedTableByID(ctx.EvalCtx.Ctx(), ctx.Txn, desc.ID)
+		if leased != nil && leased.GetVersion() == desc.Version {
```
Is checking the version even necessary at all?

ajwerner replied:
I believe that it is, yes. Or, at least, I certainly don't want to have to convince myself it's not. The alternative is that the descriptor is either newer or older. 

It's worth noting that much hand-wringing was done to convince ourselves that it's safe to use enums that are out of sync because the values are always there if they're going to be used. That made me uneasy at the time.  I worry about more complex cross-descriptor interactions. 

My preference in the longer term is to encode id and version pairs for all needed descriptors and then make it cheap to retrieve them in one go. Generally they'll be cached. That way we'd have nice invariants that all of the exact descriptors matched across machines. 

The edge case is if the descriptor hasn't been committed yet. In that case we'll still need to serialize it all. That fact has always made me sort of sad.

postamar replied:
Thanks for this information, I was also wondering about the enums. I agree with your general assessment.

postamar commented:
```
name                      old time/op    new time/op    delta
KV/Insert/SQL/rows=1-16      311µs ± 4%     307µs ±12%     ~     (p=1.000 n=9+9)
KV/Insert/SQL/rows=10-16     380µs ±11%     399µs ± 7%     ~     (p=0.083 n=10+8)
KV/Update/SQL/rows=1-16      398µs ±10%     401µs ±10%     ~     (p=0.780 n=10+9)
KV/Update/SQL/rows=10-16     628µs ±12%     636µs ± 6%     ~     (p=0.842 n=10+9)
KV/Delete/SQL/rows=1-16      287µs ±11%     288µs ± 6%     ~     (p=1.000 n=10+9)
KV/Delete/SQL/rows=10-16     399µs ± 7%     408µs ± 8%     ~     (p=0.167 n=8+9)
KV/Scan/SQL/rows=1-16        221µs ± 5%     216µs ±10%     ~     (p=0.447 n=9+10)
KV/Scan/SQL/rows=10-16       223µs ±12%     226µs ±14%     ~     (p=0.971 n=10+10)

name                      old alloc/op   new alloc/op   delta
KV/Scan/SQL/rows=1-16       19.5kB ± 1%    17.3kB ± 1%  -11.12%  (p=0.000 n=10+10)
KV/Scan/SQL/rows=10-16      21.1kB ± 0%    19.0kB ± 0%  -10.21%  (p=0.000 n=10+8)
KV/Update/SQL/rows=1-16     45.9kB ± 1%    43.9kB ± 1%   -4.46%  (p=0.000 n=10+9)
KV/Update/SQL/rows=10-16     105kB ± 1%     103kB ± 2%   -2.03%  (p=0.000 n=10+9)
KV/Insert/SQL/rows=1-16     37.7kB ± 1%    37.7kB ± 1%     ~     (p=0.796 n=9+9)
KV/Insert/SQL/rows=10-16    76.9kB ± 0%    76.8kB ± 0%     ~     (p=0.720 n=10+9)
KV/Delete/SQL/rows=1-16     41.2kB ± 0%    41.2kB ± 0%     ~     (p=0.888 n=9+8)
KV/Delete/SQL/rows=10-16    64.5kB ± 2%    64.5kB ± 1%     ~     (p=0.905 n=10+9)

name                      old allocs/op  new allocs/op  delta
KV/Scan/SQL/rows=1-16          222 ± 0%       210 ± 1%   -5.37%  (p=0.000 n=10+10)
KV/Scan/SQL/rows=10-16         257 ± 0%       245 ± 0%   -4.71%  (p=0.000 n=10+10)
KV/Update/SQL/rows=1-16        452 ± 0%       441 ± 0%   -2.52%  (p=0.000 n=10+6)
KV/Update/SQL/rows=10-16       755 ± 0%       742 ± 0%   -1.69%  (p=0.000 n=8+8)
KV/Delete/SQL/rows=1-16        293 ± 0%       292 ± 0%   -0.27%  (p=0.002 n=10+6)
KV/Insert/SQL/rows=1-16        290 ± 1%       289 ± 0%     ~     (p=0.172 n=9+9)
KV/Insert/SQL/rows=10-16       489 ± 0%       488 ± 0%     ~     (p=0.101 n=10+9)
KV/Delete/SQL/rows=10-16       474 ± 0%       473 ± 0%     ~     (p=0.189 n=10+7)
```

ajwerner reviewed (approved):
Nice

yuzefovich reviewed (commented):
Awesome! :rocket: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74722#-:-MtC0y7l1aoOIUZzpGgW:bycfel3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, and @yuzefovich)

___
*[pkg/sql/colfetcher/colbatch_scan.go, line 203 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74722#-MtC19QiDmrBuI-_FZOI:-MtC19QjCq_eIkgSOw5J:b-m0c1u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cefc87c9406c958d824f96482d7c2230e03b85f9/pkg/sql/colfetcher/colbatch_scan.go#L203)):*
> ```Go
> 
> 	limitHint := rowinfra.RowLimit(execinfra.LimitHint(spec.LimitHint, post))
> 	// TODO(ajwerner): The need to construct an immutable here
> ```

This TODO can now be removed, right?
___
*[pkg/sql/colfetcher/index_join.go, line 409 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74722#-MtC1DD3BqoDY1bsuXjc:-MtC1DD3BqoDY1bsuXjd:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/cefc87c9406c958d824f96482d7c2230e03b85f9/pkg/sql/colfetcher/index_join.go#L409)):*
> ```Go
> 	}
> 
> 	// TODO(ajwerner): The need to construct an immutable here
> ```

ditto


<!-- Sent from Reviewable.io -->

postamar commented:
Yes, good catch on those TODOs, they flew under my nose. I'm still messing around with this, btw, and I'll keep the microbenchmark results updated.

postamar commented:
I made only very minor changes since the reviews: removed the TODOs and removed an unnecessary cast that the linter was complaining about. I'll merge as soon as this passes CI.

This change now opens the door to replacing the whole table descriptor proto in the specs with an id+version pair. This ought to have quite an effect in reducing the footprint of proto de/serialization! This requires some version gating and improvements to the catalog API. The latter (batching + retrieval by version in addition to by ID) is stuff that Schema needs to do anyway at some point.

ajwerner reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> this is a big win on its own

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74722#-:-MtEDljn2EwYan4zN6kB:b-lx6vbd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, and @yuzefovich)



<!-- Sent from Reviewable.io -->

postamar commented:
Thanks for the reviews!

bors r+

nvanbenschoten commented:
Very nice! Does this resolve https://github.com/cockroachdb/cockroach/issues/74352 on its own, or would you also like to address the "replacing the whole table descriptor proto in the specs with an id+version pair" point first?

postamar commented:
Thanks! No, I think it only resolves part of it. Indeed, the other part would be not shipping around the whole table descriptor proto in the first place, which is a lot more work than I'm prepared to do right now.
